src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs

```diff
@@ -664,6 +659,33 @@ public ResultWrapper<IEnumerable<FilterLog>> eth_getLogs(Filter filter)
         {
             return GetFailureResult<IEnumerable<FilterLog>>(exception, _ethSyncingInfo.SyncMode.HaveNotSyncedReceiptsYet());
         }
+
+        ResultWrapper<IEnumerable<FilterLog>> FailWithNoHeadersSyncedYet(SearchResult<BlockHeader> blockResult)
+            => GetFailureResult<IEnumerable<FilterLog>, BlockHeader>(blockResult, _ethSyncingInfo.SyncMode.HaveNotSyncedHeadersYet());
+
+        static bool TryFindOrUseLatest(IBlockFinder blockFinder, ref BlockParameter blockParameter, out SearchResult<BlockHeader> blockResult, out long? sourceBlockNumber)
+        {
+            blockResult = blockFinder.SearchForHeader(blockParameter);
+
+            if (blockResult.IsError)
+            {
+                sourceBlockNumber = blockParameter.Type is BlockParameterType.BlockNumber ? blockParameter.BlockNumber.Value : null;
+
+                if (blockParameter.Type is BlockParameterType.BlockNumber &&
+                    blockFinder.Head?.Number is not null &&
+                    blockFinder.Head.Number < blockParameter.BlockNumber)
```

Comment:
`blockParameter.Type is BlockParameterType.BlockNumber` check is done twice.

I am not sure if I understand intention of that `if` condition. Can you make it clearer?

Reply:
It's not optimal in rare corner cases, but is surprisingly less complicated

Reply:
Can you change the code so the intentions are more readable?